File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        rule= {
            ']':'[',
            '}':'{',
            ')':'('
        }
        stack=[]
        for c in s:
            if c in rule:
                if stack and stack[-1] == rule[c]: 
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return True if not stack else False
        

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack= new Stack<>();
        for(char c:s.toCharArray()){
            if(c=='('){
                stack.add(')');
            }
            else if(c=='{'){
                stack.add('}');
            }
            else if(c=='['){
                stack.add(']');
            }
            else{
                if(stack.isEmpty()||c!=stack.peek()){
                    return false;
                }else{
                    stack.pop();
                    continue;
                }
            }
        }
        return stack.isEmpty();
        
    }
}